File: Java/Eta/ValueAdd/src/test/java/com/rtsdk/eta/valueadd/reactor/TestReactorComponent.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright (C) 2019 Refinitiv. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.rtsdk.eta.valueadd.reactor;

import static org.junit.Assert.*;

import com.rtsdk.eta.codec.Codec;
import com.rtsdk.eta.codec.Msg;
import com.rtsdk.eta.transport.BindOptions;
import com.rtsdk.eta.transport.ConnectionTypes;
import com.rtsdk.eta.transport.Server;
import com.rtsdk.eta.transport.Transport;
import com.rtsdk.eta.transport.TransportFactory;
import com.rtsdk.eta.transport.TransportReturnCodes;
import com.rtsdk.eta.valueadd.domainrep.rdm.MsgBase;

/** A component represents a consumer, provider, etc. on the network (note the Consumer and Provider subclasses). */

public abstract class TestReactorComponent {
	
    /** Reactor channel associated with this component, if connected. */
	ReactorChannel _reactorChannel;
	
	/** Whether the reactor channel associated with this component is up. */
	boolean _reactorChannelIsUp;
	
	/** ReactorRole associated with this component. */
	ReactorRole _reactorRole;
	
	/** Server associated with this component, if any. */
	Server _server;
	
	/** Reusable ReactorErrorInfo. */
	ReactorErrorInfo _errorInfo;
	
	/** Reactor associated with this component. */
	TestReactor _testReactor;
    
    int _defaultSessionLoginStreamId;
    boolean _defaultSessionLoginStreamIdIsSet;
    int _defaultSessionDirectoryStreamId;
    boolean _defaultSessionDirectoryStreamIdIsSet;
    
    static int _portToBind = 16123; /** A port to use when binding servers. Incremented with each bind. */
    static int _portToBindForReconnectTest = 16000; /** A port to use when binding servers. Incremented with each bind. */

    /** Returns the port the next bind call will use. Useful for testing reconnection
     * to servers that won't be bound until later in a test. */
    static int nextReservedServerPort()
    {
        return _portToBindForReconnectTest;
    }

    /** Returns the port of the component's server. */
    int serverPort()
    {
        assertNotNull(_server);
        return _server.portNumber();
    }
	
	TestReactor testReactor()
	{
		return _testReactor;
	}
	
	void testReactor(TestReactor testReactor)
	{
		_testReactor = testReactor;
	}
	
	ReactorChannel channel()
	{
		return _reactorChannel;
	}
		
	protected TestReactorComponent(TestReactor testReactor)
	{

		_errorInfo = ReactorFactory.createReactorErrorInfo();
		_testReactor = testReactor;
		_testReactor.addComponent(this);
	}
	
	public ReactorRole reactorRole()
	{
		return _reactorRole;
	}
	
	public ReactorChannel reactorChannel()
	{
		return _reactorChannel;
	}
	
	public void reactorChannel(ReactorChannel reactorChannel)
	{
		_reactorChannel = reactorChannel;
	}
	
	public boolean reactorChannelIsUp()
	{
		return _reactorChannelIsUp;
	}
	
	public void reactorChannelIsUp(boolean reactorChannelIsUp)
	{
		_reactorChannelIsUp = reactorChannelIsUp;
	}
	
	/** Stores the login stream ID for this component. Used if a login stream is automatically setup as part of opening a session. */
	public void defaultSessionLoginStreamId(int defaultSessionLoginStreamId)
	{
		_defaultSessionLoginStreamIdIsSet = true;
		_defaultSessionLoginStreamId = defaultSessionLoginStreamId;
		_defaultSessionLoginStreamIdIsSet = true;
	}
	
	/** If a login stream was automatically opened as part of opening a session, returns the ID of that stream for this component. */
	public int defaultSessionLoginStreamId()
	{
		assertTrue(_defaultSessionLoginStreamIdIsSet);
		return _defaultSessionLoginStreamId;
	}
	
	/** Stores the directory stream ID for this component. Used if a directory stream is automatically setup as part of opening a session. */
	public void defaultSessionDirectoryStreamId(int defaultSessionDirectoryStreamId)
	{
		_defaultSessionDirectoryStreamIdIsSet = true;
		_defaultSessionDirectoryStreamId = defaultSessionDirectoryStreamId;
		_defaultSessionDirectoryStreamIdIsSet = true;
	}

	/** If a directory stream was automatically opened as part of opening a session, returns the ID of that stream for this component. */
	public int defaultSessionDirectoryStreamId()
	{
		assertTrue(_defaultSessionDirectoryStreamIdIsSet);
		return _defaultSessionDirectoryStreamId;
	}
	
	Server server()
	{
		return _server;
	}
	
	public void bindForReconnectTest(ConsumerProviderSessionOptions opts)
	{
        if (opts.connectionType() != ConnectionTypes.RELIABLE_MCAST)
        {
            BindOptions bindOpts = TransportFactory.createBindOptions();
            bindOpts.clear();
            bindOpts.majorVersion(Codec.majorVersion());
            bindOpts.minorVersion(Codec.minorVersion());
            bindOpts.serviceName(String.valueOf(_portToBindForReconnectTest++));
            bindOpts.pingTimeout(opts.pingTimeout());
            bindOpts.minPingTimeout(opts.pingTimeout());
            _server = Transport.bind(bindOpts, _errorInfo.error());
            assertNotNull("bind failed: " + _errorInfo.error().errorId() + "(" + _errorInfo.error().text() + ")",
                _server);
        }

        _testReactor.registerComponentServer(this);
	}

	public void bind(ConsumerProviderSessionOptions opts)
	{
        if (opts.connectionType() != ConnectionTypes.RELIABLE_MCAST)
        {
        	BindOptions bindOpts = TransportFactory.createBindOptions();
            bindOpts.clear();
            bindOpts.majorVersion(Codec.majorVersion());
            bindOpts.minorVersion(Codec.minorVersion());
            bindOpts.pingTimeout(opts.pingTimeout());
            bindOpts.minPingTimeout(opts.pingTimeout());
            /* allow multiple tests to run at the same time, if the port is in use it might mean that
               another parallel test is using this port, so just try to get another port */
            while (_server == null)
            {
                bindOpts.serviceName(String.valueOf(_portToBind++));
                _server = Transport.bind(bindOpts, _errorInfo.error());
            }
        }
        
        _testReactor.registerComponentServer(this);
	}	
	
    /** Sends a Msg to the component's channel. */
    int submit(Msg msg, ReactorSubmitOptions submitOptions)
    {
    	return submit(msg, submitOptions, false);
    }

    int submit(Msg msg, ReactorSubmitOptions submitOptions, boolean expectFailure)
    {
        int ret;

        ret = _reactorChannel.submit(msg, submitOptions, _errorInfo);
        
        if (!expectFailure) {
            assertTrue("submit failed: " + ret + "(" + _errorInfo.location() + "--" + _errorInfo.error().text() + ")",
                    ret >= ReactorReturnCodes.SUCCESS);
        }

        return ret;
    }
    
    /** Sends a Msg to the component's channel, and dispatches to ensure no events are received and any internal flush events are processed. */
    int submitAndDispatch(Msg msg, ReactorSubmitOptions submitOptions)
    {
        return submitAndDispatch(msg, submitOptions, false);
    }

    /** Sends a Msg to the component's channel, and dispatches. Allowing any unprocessed messages left*/
    int submitAndDispatch(Msg msg, ReactorSubmitOptions submitOptions, boolean expectFailures)
    {
        int ret = submit(msg, submitOptions, expectFailures);
        testReactor().dispatch(expectFailures ? -1 : 0);
        return ret;
    }
    

    /** Sends an RDM message to the component's channel. */
    int submit(MsgBase msg, ReactorSubmitOptions submitOptions)
    {
        int ret;

        ret = _reactorChannel.submit(msg, submitOptions, _errorInfo);
        
        assertTrue("submit failed: " + ret + "(" + _errorInfo.location() + "--" + _errorInfo.error().text() + ")",
                ret >= ReactorReturnCodes.SUCCESS);
        
        return ret;
    }
    
	/** Sends an RDM message to the component's channel, and dispatches to ensure no events are received and any internal flush events are processed. */
    int submitAndDispatch(MsgBase msg, ReactorSubmitOptions submitOptions)
    {
        int ret = submit(msg, submitOptions);
        testReactor().dispatch(0);
        return ret;
    }
	
	/** Disconnect a consumer and provider component and clean them up. */
	public static void closeSession(Consumer consumer, Provider provider)
	{
		closeSession(consumer,provider,false);
	}
	/** Disconnect a consumer and provider component and clean them up.
	 * Do additional checks to not fail on dirty client disonnection.
	 * */
	public static void closeSession(Consumer consumer, Provider provider, boolean expectConsumerFailure)
	{
		/* Make sure there's nothing left in the dispatch queue. */
		consumer.testReactor().dispatch(0, expectConsumerFailure);
		provider.testReactor().dispatch(0);
		
		consumer.close();
		provider.close();
	}

    /** Closes the component's channel. */
    void closeChannel()
    {
        if(ReactorChannel.State.CLOSED!=_reactorChannel.state()){
            assertEquals(ReactorReturnCodes.SUCCESS, _reactorChannel.close(_errorInfo));
        }
        _reactorChannelIsUp = false;
        _reactorChannel = null;
    }
	
	/** Close a component and remove it from its associated TestReactor. 
     * Closes any associated server and reactor channel. */
	void close()
	{
        assertNotNull(_testReactor);
		if (_server != null)
		{
			assertEquals(TransportReturnCodes.SUCCESS, _server.close(_errorInfo.error()));
			_server = null;
		}
		
		if (_reactorChannel != null)
            closeChannel();

        _testReactor.removeComponent(this);
        _testReactor = null;
	}
}

File: CSharp/Eta/Src/Tests/ValueAddTest/TestReactorComponent.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using Xunit;

using LSEG.Eta.Codec;
using LSEG.Eta.Transports;
using LSEG.Eta.ValueAdd.Rdm;
using LSEG.Eta.ValueAdd.Reactor;

namespace LSEG.Eta.ValuedAdd.Tests;

/** A component represents a consumer, provider, etc. on the network (note the Consumer and Provider subclasses). */
public abstract class TestReactorComponent
{
    /// <summary>
    /// Reactor channel associated with this component, if connected.
    /// </summary>
    protected ReactorChannel m_ReactorChannel;

    /// <summary>
    /// Whether the reactor channel associated with this component is up.
    /// </summary>
    protected bool m_ReactorChannelIsUp;

    /// <summary>
    /// ReactorRole associated with this component.
    /// </summary>
    public ReactorRole ReactorRole { get; set; }

    /// <summary>
    /// Server associated with this component, if any.
    /// </summary>
    public IServer Server { get; set; }

    /// <summary>
    /// Returns the port of the component's server.
    /// </summary>
    public int ServerPort
    {
        get
        {
            Assert.NotNull(Server);
            return Server.PortNumber;
        }
    }

    /// <summary>
    /// Reactor associated with this component.
    /// </summary>
    public TestReactor TestReactor { get; set; }

    int m_DefaultSessionLoginStreamId;
    bool m_DefaultSessionLoginStreamIdIsSet;
    int m_DefaultSessionDirectoryStreamId;
    bool m_DefaultSessionDirectoryStreamIdIsSet;

    /// <summary>
    /// A port to use when binding servers. Incremented with each bind.
    /// </summary>
    static int m_PortToBind = 16123;

    /// <summary>
    /// A port to use when binding servers. Incremented with each bind.
    /// </summary>
    static int m_PortToBindForReconnectTest = 16000;

    /** Returns the port the next bind call will use. Useful for testing reconnection
     * to servers that won't be bound until later in a test. */
    static int nextReservedServerPort()
    {
        return m_PortToBindForReconnectTest;
    }

    protected TestReactorComponent(TestReactor testReactor)
    {
        TestReactor = testReactor;
        TestReactor.AddComponent(this);
    }

    public ReactorChannel ReactorChannel
    {
        get => m_ReactorChannel;
        set
        {
            m_ReactorChannel = value;
        }
    }


    public bool IsReactorChannelUp
    {
        get => m_ReactorChannelIsUp;
        set
        {
            m_ReactorChannelIsUp = value;
        }
    }

    /// <summary>
    /// Stores the login stream ID for this component.
    /// </summary>
    ///
    /// <remarks>
    /// Used if a login stream is automatically setup as part of opening a session.
    /// If a login stream was automatically opened as part of opening a session,
    /// returns the ID of that stream for this component.
    /// </remarks>
    ///
    public int DefaultSessionLoginStreamId
    {
        get
        {
            Assert.True(m_DefaultSessionLoginStreamIdIsSet);
            return m_DefaultSessionLoginStreamId;
        }
        set
        {
            m_DefaultSessionLoginStreamIdIsSet = true;
            m_DefaultSessionLoginStreamId = value;
        }
    }

    /// <summary>
    /// If a directory stream was automatically opened as part of
    /// opening a session, returns the ID of that stream for this component.
    /// </summary>
    ///
    /// <remarks>
    /// Stores the directory stream ID for this component. Used if a directory
    /// stream is automatically setup as part of opening a session.
    /// </remarks>
    ///
    public int DefaultSessionDirectoryStreamId
    {
        get
        {
            Assert.True(m_DefaultSessionDirectoryStreamIdIsSet);
            return m_DefaultSessionDirectoryStreamId;
        }

        set
        {
            m_DefaultSessionDirectoryStreamIdIsSet = true;
            m_DefaultSessionDirectoryStreamId = value;
        }
    }

    public void BindForReconnectTest(ConsumerProviderSessionOptions opts)
    {
        BindOptions bindOpts = new();
        bindOpts.Clear();
        bindOpts.MajorVersion = Codec.Codec.MajorVersion();
        bindOpts.MinorVersion = Codec.Codec.MinorVersion();
        bindOpts.ServiceName = (m_PortToBindForReconnectTest++).ToString();
        bindOpts.PingTimeout = (int)opts.PingTimeout.TotalSeconds;
        bindOpts.MinPingTimeout = (int)opts.PingTimeout.TotalSeconds;
        Server = Transport.Bind(bindOpts, out var bindError);
        Assert.True(Server != null,
                    $"bind failed: {bindError.ErrorId} ({bindError.Text})");

        TestReactor.RegisterComponentServer(this);
    }


    public void Bind(ConsumerProviderSessionOptions opts)
    {
        BindOptions bindOpts = new();
        bindOpts.Clear();
        bindOpts.MajorVersion = Codec.Codec.MajorVersion();
        bindOpts.MinorVersion = Codec.Codec.MinorVersion();
        bindOpts.PingTimeout = (int)opts.PingTimeout.TotalSeconds;
        bindOpts.MinPingTimeout = (int)opts.PingTimeout.TotalSeconds;
        bindOpts.GuaranteedOutputBuffers = opts.NumOfGuaranteedBuffers;
        bindOpts.CompressionType = opts.CompressionType;

        // allow multiple tests to run at the same time, if the port is in use it might mean that
        // another parallel test is using this port, so just try to get another port
        while (Server == null)
        {
            bindOpts.ServiceName = (m_PortToBind++).ToString();
            Server = Transport.Bind(bindOpts, out _);
        }

        TestReactor.RegisterComponentServer(this);
    }

    /// <summary>
    /// Sends a Msg to the component's channel.
    /// </summary>
    ///
    public ReactorReturnCode Submit(Msg msg, ReactorSubmitOptions submitOptions)
    {
        return Submit(msg, submitOptions, false);
    }


    public ReactorReturnCode Submit(Msg msg, ReactorSubmitOptions submitOptions, bool expectFailure)
    {
        ReactorReturnCode ret = m_ReactorChannel.Submit(msg, submitOptions, out var errorInfo);

        if (!expectFailure)
        {
            Assert.True(ret >= ReactorReturnCode.SUCCESS,
                $"submit failed: {ret} ({errorInfo?.Location} -- {errorInfo?.Error?.Text})");
        }

        return ret;
    }


    /// <summary>
    /// Sends a Msg to the component's channel, and dispatches to ensure no events are
    /// received and any internal flush events are processed.
    /// </summary>
    ///
    public ReactorReturnCode SubmitAndDispatch(Msg msg, ReactorSubmitOptions submitOptions)
    {
        return SubmitAndDispatch(msg, submitOptions, false);
    }


    /// <summary>
    /// Sends a Msg to the component's channel, and dispatches. Allowing any unprocessed messages left
    /// </summary>
    public ReactorReturnCode SubmitAndDispatch(Msg msg, ReactorSubmitOptions submitOptions, bool expectFailures)
    {
        ReactorReturnCode ret = Submit(msg, submitOptions, expectFailures);
        TestReactor.Dispatch(expectFailures ? -1 : 0);
        return ret;
    }


    /// <summary>
    /// Sends an RDM message to the component's channel.
    /// </summary>
    public ReactorReturnCode Submit(MsgBase msg, ReactorSubmitOptions submitOptions)
    {
        ReactorReturnCode ret = m_ReactorChannel.Submit(msg, submitOptions, out var errorInfo);

        Assert.True(ret >= ReactorReturnCode.SUCCESS,
            $"submit failed: {ret} ({errorInfo?.Location} -- {errorInfo?.Error?.Text})");

        return ret;
    }


    /** Sends an RDM message to the component's channel, and dispatches to ensure no
     * events are received and any internal flush events are processed. */
    public ReactorReturnCode SubmitAndDispatch(MsgBase msg, ReactorSubmitOptions submitOptions)
    {
        ReactorReturnCode ret = Submit(msg, submitOptions);
        TestReactor.Dispatch(0);
        return ret;
    }

    /// <summary>
    /// Disconnect a consumer and provider component and clean them up.
    /// </summary>
    public static void CloseSession(Consumer consumer, Provider provider)
    {
        CloseSession(consumer, provider, false);
    }


    /// <summary>
    /// Disconnect a consumer and provider component and clean them up.
    /// Do additional checks to not fail on dirty client disonnection.
    /// </summary>
    public static void CloseSession(Consumer consumer, Provider provider, bool expectConsumerFailure)
    {
        /* Make sure there's nothing left in the dispatch queue. */
        consumer.TestReactor.Dispatch(0, expectConsumerFailure);
        provider.TestReactor.Dispatch(0);

        consumer.Close();
        provider.Close();
    }

    /// <summary>
    /// Closes the component's channel.
    /// </summary>
    internal void CloseChannel()
    {
        if (ReactorChannelState.CLOSED != m_ReactorChannel.State)
        {
            Assert.Equal(ReactorReturnCode.SUCCESS, m_ReactorChannel.Close(out _));
        }
        m_ReactorChannelIsUp = false;
        m_ReactorChannel = null;
    }


    /// <summary>
    /// Close a component and remove it from its associated TestReactor.
    /// Closes any associated server and reactor channel.
    /// </summary>
    public void Close()
    {
        Assert.NotNull(TestReactor);
        if (Server != null)
        {
            Assert.Equal(TransportReturnCode.SUCCESS, Server.Close(out _));
            Server = null;
        }

        if (m_ReactorChannel != null)
            CloseChannel();

        TestReactor.RemoveComponent(this);
        TestReactor = null;
    }

    public ReactorReturnCode GetChannelInfo(ReactorChannelInfo reactorChannelInfo)
    {
        if(m_ReactorChannelIsUp && m_ReactorChannel != null)
        {
            return m_ReactorChannel.Info(reactorChannelInfo, out _);
        }
        else
        {
            return ReactorReturnCode.FAILURE;
        }
    }
}
